antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/misc/Array2DHashSet.kt

```diff
@@ -9,484 +9,539 @@ package org.antlr.v4.kotlinruntime.misc
+      return true
     }
 
-    protected fun getOrAddImpl(o: T): T {
-        val b = getBucket(o)
-        var bucket: Array<T>? = buckets[b]
+    if (other !is Array2DHashSet<*>) {
+      return false
+    }
 
-        // NEW BUCKET
-        if (bucket == null) {
-            bucket = createBucket(initialBucketCapacity) as Array<T>
-            bucket[0] = o
-            buckets[b] = bucket
-            n++
-            return o
-        }
+    return other.size == size && containsAll(other)
+  }
 
-        // LOOK FOR IT IN BUCKET
-        for (i in bucket.indices) {
-            val existing = bucket[i]
-            if (existing == null) { // empty slot; not there, add.
-                bucket[i] = o
-                n++
-                return o
-            }
-            if (comparator.equals(existing, o)) return existing // found existing, quit
-        }
+  protected fun expand() {
+    val old = buckets
+    currentPrime += 4
 
-        // FULL BUCKET, expand and add to end
-        val oldLength = bucket.size
-        bucket = bucket.copyOf(bucket.size * 2) as Array<T>
-        buckets[b] = bucket
-        bucket[oldLength] = o // add to end
-        n++
-        return o
-    }
+    val newCapacity = buckets.size * 2
+    val newTable = createBuckets(newCapacity)
+    val newBucketLengths = IntArray(newTable.size)
+
+    buckets = newTable
+    threshold = (newCapacity * LOAD_FACTOR).toInt()
+
+    // Rehash all existing entries
+    val oldSize = size
+
+    for (bucket in old) {
+      if (bucket == null) {
+        continue
+      }
+
+      for (o in bucket) {
+        if (o == null) {
+          break
+        }
 
-    operator fun get(o: T?): T? {
-        if (o == null) return o
         val b = getBucket(o)
-        val bucket = buckets[b] ?: return null
-// no bucket
-        for (e in bucket) {
-            if (e == null) return null // empty slot; not there
-            if (comparator.equals(e, o)) return e
+        val bucketLength = newBucketLengths[b]
+        var newBucket: Array<T?>
+
+        if (bucketLength == 0) {
+          // New bucket
+          newBucket = createBucket(initialBucketCapacity)
+          newTable[b] = newBucket
+        } else {
+          newBucket = newTable[b]!!
+
+          if (bucketLength == newBucket.size) {
+            // Expand
+            newBucket = newBucket.copyOf(newBucket.size * 2)
+            newTable[b] = newBucket
+          }
         }
-        return null
+
+        newBucket[bucketLength] = o
+        newBucketLengths[b]++
+      }
     }
 
-    protected fun getBucket(o: T): Int {
-        val hash = comparator.hashCode(o)
-        return hash and (buckets.size - 1)
+    assert(n == oldSize)
+  }
+
+  override fun add(element: T): Boolean {
+    val existing = getOrAdd(element)
+    return existing === element
+  }
+
+  override val size: Int
+    get() = n
+
+  override fun isEmpty(): Boolean =
+    n == 0
+
+  // TODO(Edoardo): why is this one commented out?
```

Comment:
TODO that we should look into before merging

Reply:
Answer is we already have the typed version.
```kotlin
override fun contains(element: T): Boolean =
  containsFast(element)
```

So I've simply removed it now.